### DISCUSSION TITLE
Non-standard Primary Key in SQLAlchemy User Model

### DISCUSSION THREAD

**Discussion by Xevion** [fastapi-users/fastapi-users#1070] [Q&A] [answered]
Is there any easy way of integrating a non-standard primary key for the user model?

For example, I use `bigint`, a non-incrementing 64bit signed integer for my user model. This is rather standard, but for `fastapi-users`, presents a problem; I have very little control over the ID. I need control over this ID as `fastapi-users` does not perform any generation of the id, which is fine for `serial` or any other auto-incrementing/generating primary key.

Initially, I assumed I had no control, but I am able to set it by default using SQLAlchemy `Column`'s `default` kwarg. Here, I can pass a function that generates the default value, and thus, I can control what `fastapi-users` will commit to the database in the end.

Are there any alternatives to controlling what values get generated for the User?

The only alternative I identified was creating a custom registration route and using the [Creating a User Progmatically][creating-a-user] example. This could let me perform more checks and use more advanced logic.

As an example of why I might want more control: What about a column that is generated based on the value of the ID? I use base64 encoded identifiers throughout my app with their values embedded in URL slugs. I don't currently know of a way of doing this with default, so I can only assume I will be required to create a registration route just so I can have the slug column properly generated.

```python
class User(SQLAlchemyBaseUserTable[int], Base):
    __tablename__ = "user"

    id = sa.Column(sa.BigInteger, primary_key=True, autoincrement=False,
                   default=lambda: secrets.randbits(64) - (2 ** 63))
    slug = sa.Column(sa.String(11), nullable=False)
```

I could convert it on each request, but this is of course very inefficient.

[creating-a-user]: https://fastapi-users.github.io/fastapi-users/latest/cookbook/create-user-programmatically/

**Comment by frankie567** [comment, marked as answer]
Well, FastAPI Users lets you indeed define any type as ID, but yes, if you need more advanced logic, you'll probably need to tinker a bit. I think your best bet is to overload the `create` method of `UserManager` and have a custom `UserCreateInternal` model:

```py
class UserCreateInternal(UserCreate):
    id: int
    slug: str


class UserManager(UUIDIDMixin, BaseUserManager[User, int]):
    # ...
    async def create(
        self,
        user_create: UserCreate,
        safe: bool = False,
        request: Optional[Request] = None,
    ) -> User:
        user_id = generate_id()  # Your method for generating ID
        user_slug = generate_slug(user_id)
        user_create_internal = UserCreateInternal(**user_create.dict(), id=user_id, slug=user_slug)
        return await super().create(user_create_internal, safe=safe, request=request)
```

The good thing is that you can inject dependencies in the `get_user_manager` constructor, so if you need external services for your ID generation, you can add them.

That said, IMHO, your ID generation looks *really* complex. If you need safe, non-colliding ID generation that you can directly use in URL, UUID seem a much straightforward choice.

**Reply by Xevion** [reply]
Cool, so there is a way of actually modifying the internal logic here. Currently, I'm sticking with a safe and simple `default` kwarg with SQLAlchemy, but I've been worried about whether I should stick with `fastapi-users` in particular because so much logic is hidden from me...

By the way, I don't think there's much complexity with my ID generation. It looks a little funky, but it's dealing with conversion from a _unsigned_ 64 bit integer to a *signed* 64 bit integer. By subtracting half it's range (2^64 / 2, or 2^63), the id generated is signed and properly centered on 0.

And yes, I have studied UUID and various other non-colliding methods and have concluded that int64 is safe for me. It's just unfortunate that Postgresql does not have native support for an unsigned, non-autoincrementing 64bit integer. The way I've decided to do it is combine a slug with a base64 encoded 64bit integer. This should give me collision-proof ID generation up to 4 billion (way beyond my needs), and in the case that I ever need to scale up, I can always migrate to 128 bit, which is pretty much up there with UUID. Also, I like my URLs nice and short. UUID is waaaay too long for me.

**Reply by Xevion** [reply]
I'm adding on to my last comment about the ID generation; this has nothing to do with the original question, and I'd like feedback from an experienced developed to confirm my thoughts, but if you don't want to, I won't mind being ignored.

The reason I'm going through this measure is to make my URLs super friendly and short, but also very random so as to increase the security of my site. 32 bits is too small, 128 bits is too large, but 64 bits is right in the center. As long as you're not indexing a billion things (I am not), your site should never collide - and if it does, given that the database is accessed from the backend, you're already performing the checks. Retries can be done easily, afaik.

So, I chose 64bit encoded in base64 as compared to UUIDs, they take up way less space and are just as fast to encode/decode. UUIDs are hexadecimal only, and while they can be encoded in base64, they're still super long like that (22 characters, vs just 11 for 64bit). So, I chose 64. In the end, I use `python-slugify` with a 20 char limit, append a dash and then the encoded ID, and all my routes are formatted like that. Some models don't use 64bit, but it's only things that shouldn't be exposed to users in the background.

My routes look like this: `example.com/gallery/gallery-test-name-J0vp6pRxJBM/` for a gallery named "Gallery Test Name" with the id `1379352270292470567`. This makes my ID random and unpredictable (not linear), encodes the name of the object being accessed, can be renamed without trouble (only the last 11 characters are used to search), and is very friendly for users compared to UUIDs.

Do tell me if I am getting anything wrong here. I've already talked to one of my managers about this and they found nothing wrong, but they don't traditionally work on public web applications.

**Reply by frankie567** [reply]
> but I've been worried about whether I should stick with fastapi-users in particular because so much logic is hidden from me...

It's up to you to take that decision, and a trade-off to make: on one side, you gain lot of battle-tested logic for free, but on the other side you give up on some control and might be in trouble if your use-case doesn't fit. For the record, the FastAPI tutorial explains quite well how to implement a basic auth system: https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/

> Do tell me if I am getting anything wrong here. 

Well, there is nothing technically wrong with this approach, you seem to have thought about it thoroughly, and it'll definitely work.

It's just that, personally, I like to keep things very simple 😄 Be sure to document your ID generation algorithm in details, so your future colleagues don't get lost when they see this!

Another approach that could help you, in particular for setting the slug: [SQLAlchemy attribute events](https://docs.sqlalchemy.org/en/14/orm/events.html#attribute-events). It allows you to trigger logic when a field is set.
